### ISSUE TITLE
Mock external services in tests

### ISSUE THREAD

**Issue by dstillman** [zotero/zotero#699]
Our Travis tests are currently failing [all the time](https://travis-ci.org/zotero/zotero/builds) for reasons that have nothing to do with the code under test. Right now, for example, [we're failing](https://travis-ci.org/zotero/zotero/jobs/58499872) because of a mismatch in a title returned from an ISBN lookup. That was caused by [a translator issue](https://github.com/zotero/translators/commit/81fd1c9a99ef30b51603261db95698adc4aa5661), and it's at best distracting and at worst unnecessarily alarming to have it showing up on random commits and pull requests. I'd say we want to avoid external services almost entirely in Travis and just have them for local test runs, which the main developers should be doing all the time. In my view, Travis tests should almost always be passing on the main branch, with failures limited to unexpected side effects and untested pull requests.

We can create a utility function using [Sinon](http://sinonjs.org/) to mock ``Zotero.HTTP.request()`` for various things, but I'm not sure how we want to handle things that depend on translators, and particularly a case like this where it wasn't even that the LoC MARC response changed, but that a different site was being used entirely.

@simonster, @aurimasv, thoughts?

**Comment by dstillman** [comment]
I also think we should disable translator/style updates from the repository in Travis, so that failures are tied to specific commits. That means that a bug like this wouldn't show up until we bumped the translator submodule, which would be the relevant commit. (This one actually wouldn't have shown up for that, because it was specific to repository behavior, but as long as repo updates are enabled for local test runs, a test that checked which translator was used for a particular ISBN lookup would've caught this bug locally, even if the actual external response was mocked.)

**Comment by aurimasv** [comment]
I would like to keep Travis tests as close as possible to local tests (unless technically impossible, e.g. solving CAPTCHA) and as close as possible to what the actual client would do. IMO Travis tests are there to help fool-proof commits/PRs, so saying that this part of the software we'll ask the developers to test locally doesn't sound that great to me. Basically, if the tests pass, then, if Zotero were to be released from that commit, all its functions should behave as we expect them. Obviously easier said than done.

Regarding currently failing Add by Identifier tests. That is one of the key features of Zotero and it relies on translators. Furthermore, it does not specifically rely on any specific translators (we let the translator infrastructure decide which translator is supposed to be used). So, for one, I think we need to be testing this with the current set of translators that is live at the time of testing. I think the translator updates should remain. We may wish to consider running tests on zotero/translators branch to make sure that new commits to zotero/translators do not break key Zotero functionality. At some point, it would be great to extend those tests to actually test translators, at least for syntax.

Relying on external services is an issue. On one hand, we could mock the external service. On the other hand, since our core functionality depends on external services, we should be aware when external services start failing so we can fix it if needed. Ideally, I would like to just be notified if certain key external services (LoC, Google Scholar, PubMed, WorldCat) start failing. This could/should be a monitoring system outside of Travis. In either case, it seems that currently our best option is to mock external services. We should be able to anticipate all HTTP requests from Zotero tests. In the case of a change in translator used for Add by Identifier or Retrieve Metadata, I think those are the kind of failures we want to be aware of and the kind of failures we want to correct in our tests.

**Comment by dstillman** [comment]
> Regarding currently failing Add by Identifier tests. That is one of the key features of Zotero and it relies on translators.

Sure, but it doesn't make sense for Rintze's locale drop-down PR to fail because of a translator issue. If the commit he branched from was passing, his code should pass unless he introduces a bug, and we shouldn't have to review the Travis log before accepting his PR just to see if the failure was related to his code.

Dev builds pull the latest versions of submodules automatically, and they obviously have automatic updates enabled when people run them, but we bump the submodules to the latest versions before releases, so those will create appropriate test points even with automatic updates disabled.

> We may wish to consider running tests on zotero/translators branch to make sure that new commits to zotero/translators do not break key Zotero functionality. 

That's a good idea. I think we should do this, remove automatic translator updates from zotero/zotero, and mock external services. Then we'd be fully testing the correct code in all cases, and failures would be meaningful.

(We can discuss monitoring external services separately. We also might want to preserve a non-mocking option for local tests.)

**Comment by aurimasv** [comment]
>Sure, but it doesn't make sense for Rintze's locale drop-down PR to fail because of a translator issue.

I agree, but that's the case with any unrelated breakage. Basically, I think the translator update should have triggered a Travis failure and it shouldn't have been accepted until tests were fixed. So adding tests to zotero/translators should take care of this. I don't see a reason to disable translator updates if we cover this case though.

>Dev builds pull the latest versions of submodules automatically

Not sure I follow what you mean by "dev builds" here. Running Zotero from source? Or the "alpha" XPI that you build and publish?

>we bump the submodules to the latest versions before releases, so those will create appropriate test points even with automatic updates disabled.

Well, at least until we publish new translator updates. Or am I missing something here?

**Comment by aurimasv** [comment]
>>We may wish to consider running tests on zotero/translators branch to make sure that new commits to zotero/translators do not break key Zotero functionality.
>
>That's a good idea. I think we should do this

Are you going to tinker with that or do you want me to give it a go?

**Comment by dstillman** [comment]
> So adding tests to zotero/translators should take care of this. I don't see a reason to disable translator updates if we cover this case though.

Because of what I say above — that if you check out a passing commit, it should continue to work unless you yourself break something. That's the only way for test results to be meaningful.

> Not sure I follow what you mean by "dev builds" here.

The dev XPIs (though source installations would have automatic translator updates enabled too). My point was just that release XPIs have specific commits (submodule bumps) that would be tested for translator-induced breakage.

>> we bump the submodules to the latest versions before releases, so those will create appropriate test points even with automatic updates disabled.

> Well, at least until we publish new translator updates. Or am I missing something here?

But those will be covered by zotero/translator tests for the commits that actually introduced the breakage.

> Are you going to tinker with that or do you want me to give it a go?

I wouldn't be able to get to it for a couple weeks at least, so if you wanted to try, that'd be great, or maybe @simonster can.

**Comment by rmzelle** [comment]
How about expanding the build matrix, and putting the translator tests in their own build? Seems relatively easy with environment variables http://docs.travis-ci.com/user/build-configuration/#The-Build-Matrix

**Comment by dstillman** [comment]
Isn't that just worse than doing the tests from zotero/translators?

**Comment by rmzelle** [comment]
> Isn't that just worse than doing the tests from zotero/translators?

But isn't there an interdependency between translator code and Zotero code? E.g. to test certain Zotero code, you need a working translator, and vice versa. Maybe doing tests in zotero/translators makes sense because core Zotero code is generally more stable?

**Comment by aurimasv** [comment]
Code in zotero/translators master branch and code in zotero/zotero 4.0 branch is supposed to be release-ready, so both are supposed to be just as stable. The reason to do tests on zotero/translators is so we can fix any errors or adjust tests for translator pull requests

**Comment by dstillman** [comment]
> But isn't there an interdependency between translator code and Zotero code?

We'd still be running translator-related tests in zotero/zotero — we'd just be doing it with the version of the translators submodule that was actually checked in, so that failures would show up for the appropriate commit (i.e., the submodule bump). And then we'd test ongoing translator commits in zotero/translators.

**Comment by fbennett** [comment]
When the dust settles, it would be good to have a note in the source on where the translators are pulled from, and where that happens (or how it can be controlled). Late to the party, I've just been [around the rose bush](https://groups.google.com/forum/#!topic/zotero-dev/DWh_axWXi5A) on the failing translator issue, and it's all clear now - except that I still don't have a clue where the translators in the test build are coming from. I'll need to adjust that, because MLZ/Juris-M adds some legal-site translators that don't belong in official Zotero yet.

**Comment by dstillman** [comment]
The test framework just creates a new data directory each time, so it's the same as any source installation of Zotero — it pulls from the 'translators' directory of the extension directory (i.e., the source code) into the data directory, and then it auto-updates from the repo. When we resolve all of this, there'll be a command-line flag to enable and disable automatic updates from the repo, which will be enabled by default for Travis runs.

**Comment by fbennett** [comment]
Got it - it sounds like I needed to do a push.

**Comment by aurimasv** [comment]
>if you check out a passing commit, it should continue to work unless you yourself break something

OK, I agree with that, but this can only hold for local tests, not Travis testing. Travis tests against the merge commit (i.e. what would happen if PR were merged into 4.0/master/whatever), so if your PR is not against the current tip of the master branch _and_ it passes tests locally, you're still not guaranteed that Travis-CI tests will pass and, of course, that is how it's supposed to be.

So, I still think that we need to be testing against the current state of Zotero code base and I do agree that local tests should not break unless you mess something up yourself. That means, that we should disable translator updates locally, but keep them for Travis-CI testing.

The random Travis-CI failures need to actually be resolved by updating tests. That, in addition to mocking external services (which would eliminate the semi-randomness of translator-based tests) and running CI tests on zotero/translators, should make sure that we are performing proper testing and that we're not starting to fail builds randomly.
